### PR TITLE
Add process guide for migrating packages into the core monorepo

### PR DESCRIPTION
**Pull request by MajorLift**
## Explanation

This document outlines the process for migrating a MetaMask library into the core monorepo. The migration target is assumed to comply with the requirements defined by [`metamask-module-template`](https://github.com/MetaMask/metamask-module-template) and [`module-lint`](https://github.com/MetaMask/module-lint).

## References

- Closes https://github.com/MetaMask/core/issues/1832

## Changelog

N/A

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

File: docs/package-migration-process-guide.md
Comment (mcmire) [inline docs/package-migration-process-guide.md:15]: What do you think about explaining how to do this?

File: docs/package-migration-process-guide.md
Comment (mcmire) [inline docs/package-migration-process-guide.md:44]: Why not start at 1? :)

File: docs/package-migration-process-guide.md
Comment (mcmire) [inline docs/package-migration-process-guide.md:68]: Typo?

``` suggestion
- Use the `scripts/migrate-tags.sh` tool to port the source repo's release tags into the migrated git history in core, following these [instructions](https://github.com/MetaMask/core/issues/1800).
```

or maybe the reader should consult the document on migrating tags instead?

``` suggestion
- Use the `scripts/migrate-tags.sh` tool to port the source repo's release tags into the migrated git history in core, following [these instructions](./migrate-tags.md).
```

File: docs/package-migration-process-guide.md
Comment (mcmire) [inline docs/package-migration-process-guide.md:70]: I think we should highlight the use of a fork for testing. It feels like this could get missed (yes, even though `migrate-tags.md` mentions this).

What do you think about something like:

``` suggestion
- Create a fork of the core monorepo for testing.
- Push the ported tags to a fork of the core monorepo for testing. **Do not run the script against `MetaMask/core` before testing it on a fork.**

File: docs/package-migration-process-guide.md
Comment (mcmire) [inline docs/package-migration-process-guide.md:198]: We also need to add the migration target to the list of packages. Perhaps this?

``` suggestion
1. **[PR#14]** Add migration target to the list of packages in the README as well as the dependency graph in the README (the latter can be updated automatically by running `yarn generate-dependency-graph`).

File: docs/package-migration-process-guide.md
Comment (mcmire) [inline docs/package-migration-process-guide.md:157]: This step feels redundant, and possibly misleading. Merging the PR does this already, so perhaps this is implied and can be removed?

``` suggestion
```

**Review by mcmire — COMMENTED** [review]
That's a lot of PRs to have to create for one package 😅  But it all makes sense, and it's nice that many of the PRs will be small.

Thanks for putting this together. This looks really comprehensive. I just had a few suggestions.

File: docs/package-migration-process-guide.md
Comment (MajorLift) [reply docs/package-migration-process-guide.md]: Fixed here 18ccc1875e0fb33b296e847c12a5882d85c355fa

File: docs/package-migration-process-guide.md
Comment (MajorLift) [reply docs/package-migration-process-guide.md:15]: Fixed here 2cf03afa525a73e9a0e9636e27a8167eae6b8b87

**Review by mikesposito — DISMISSED** [review]
Looks great!

Nice to see that you also have a detailed _preparation_ phase to outline the steps to be performed prior to the migration, and that there are plenty of example PRs!

File: docs/package-migration-process-guide.md
Comment (mcmire) [inline docs/package-migration-process-guide.md:83]: I believe that the options need to be provided as snake-case, is that correct?

``` suggestion
- If the package has been renamed or needs to be renamed with the `@metamask/` namespace, supply two arguments to the script: `--version-before-package-rename`, `--tag-prefix-before-package-rename`.

File: docs/package-migration-process-guide.md
Comment (mcmire) [inline docs/package-migration-process-guide.md:73]: Nit: Worth using an emoji to make this pop out?

``` suggestion
#### ⚠️ Caution

- DO NOT rebase the `migrate-<package-name>` branch, as this will disrupt the git history.
- Ensure that superfluous merge commits to the main branch don't pollute the migrated git history.
  - Coordinate with the team to minimize the time that this PR stays open.
  - If necessary, replace the PR branch with a cleaned-up commit history by rerunning the git history migration steps before merging the PR.
  - For further context on this: https://github.com/MetaMask/core/pull/1804#issuecomment-1771445829
- Merge PR **without squashing** to preserve the migrated git commit history.
  - Contact a [**maintainer**](https://github.com/orgs/MetaMask/teams/engineering?query=role%3Amaintainer) to temporarily enable merge commits into main.
- [Example PR](https://github.com/MetaMask/core/pull/1872)
```

or, use a WARNING block:

``` suggestion
> [!WARNING]
> - DO NOT rebase the `migrate-<package-name>` branch, as this will disrupt the git history.
> - Ensure that superfluous merge commits to the main branch don't pollute the migrated git history.
>   - Coordinate with the team to minimize the time that this PR stays open.
>   - If necessary, replace the PR branch with a cleaned-up commit history by rerunning the git history migration steps before merging the PR.
>   - For further context on this: https://github.com/MetaMask/core/pull/1804#issuecomment-1771445829
> - Merge PR **without squashing** to preserve the migrated git commit history.
>   - Contact a [**maintainer**](https://github.com/orgs/MetaMask/teams/engineering?query=role%3Amaintainer) to temporarily enable merge commits into main.
> - [Example PR](https://github.com/MetaMask/core/pull/1872)
```

Whichever looks better to you.

**Review by mcmire — COMMENTED** [review]
One more thing but otherwise I think this is ready to go!

File: docs/package-migration-process-guide.md
Comment (MajorLift) [reply docs/package-migration-process-guide.md]: Yep you're right. Good catch!

File: docs/package-migration-process-guide.md
Comment (MajorLift) [reply docs/package-migration-process-guide.md]: Went with the second option here: https://github.com/MetaMask/core/pull/2071/commits/a25df15255039d459947cd1133201589c8533a5d

**Review by mcmire — APPROVED** [review]
Looks good!

**Pull request merged**
